command/connect/expose/expose.go

```diff
@@ -0,0 +1,236 @@
+
+	c.flags.StringVar(&c.service, "service", "",
+		"(Required) The name of destination service to expose. A namespace "+
+			"can optionally be specified as a prefix via the 'namespace/service' format.")
+
+	c.flags.IntVar(&c.port, "port", 0,
+		"(Required) The listener port to use for the service on the Ingress gateway.")
+
+	c.flags.StringVar(&c.protocol, "protocol", "tcp",
+		"The protocol for the service. Defaults to 'tcp'.")
+
+	c.flags.Var(&c.hosts, "host", "Additional DNS hostname to use for routing to this service."+
+		"Can be specified multiple times.")
+
+	c.http = &flags.HTTPFlags{}
+	flags.Merge(c.flags, c.http.ClientFlags())
+	flags.Merge(c.flags, c.http.ServerFlags())
+	c.help = flags.Usage(help, c.flags)
+}
+
+func (c *cmd) Run(args []string) int {
+	if err := c.flags.Parse(args); err != nil {
+		if err == flag.ErrHelp {
+			return 0
+		}
+		c.UI.Error(fmt.Sprintf("Failed to parse args: %v", err))
+		return 1
+	}
+
+	// Set up a client.
+	client, err := c.http.APIClient()
+	if err != nil {
+		c.UI.Error(fmt.Sprintf("Error initializing client: %s", err))
+		return 1
+	}
+
+	// Check for any missing or invalid flag values.
+	if c.service == "" {
+		c.UI.Error("A service name must be given via the -service flag.")
+		return 1
+	}
+	svc, svcNamespace, err := create.ParseIntentionTarget(c.service)
+	if err != nil {
+		c.UI.Error(fmt.Sprintf("Invalid service name: %s", err))
+		return 1
+	}
+
+	if c.ingressGateway == "" {
+		c.UI.Error("An ingress gateway service must be given via the -ingress-gateway flag.")
+		return 1
+	}
+	gateway, gatewayNamespace, err := create.ParseIntentionTarget(c.ingressGateway)
+	if err != nil {
+		c.UI.Error(fmt.Sprintf("Invalid ingress gateway name: %s", err))
+		return 1
+	}
+
+	if c.port == 0 {
+		c.UI.Error("A port must be provided via the -port flag.")
+		return 1
+	}
+
+	// First get the config entry for the ingress gateway, if it exists. Don't error if it's a 404 as that
+	// just means we'll need to create a new config entry.
+	conf, _, err := client.ConfigEntries().Get(api.IngressGateway, gateway, nil)
+	if err != nil && !strings.Contains(err.Error(), agent.ConfigEntryNotFoundErr) {
+		c.UI.Error(fmt.Sprintf("Error fetching existing ingress gateway configuration: %s", err))
+		return 1
+	}
+	if conf == nil {
+		conf = &api.IngressGatewayConfigEntry{
+			Kind:      api.IngressGateway,
+			Name:      gateway,
+			Namespace: gatewayNamespace,
+		}
+	}
+
+	// Make sure the flags don't conflict with existing config.
+	ingressConf, ok := conf.(*api.IngressGatewayConfigEntry)
+	if !ok {
+		// This should never happen
+		c.UI.Error(fmt.Sprintf("Config entry is an invalid type: %T", conf))
+		return 1
+	}
+
+	listenerIdx := -1
+	for i, listener := range ingressConf.Listeners {
+		// Make sure the service isn't already exposed in this gateway
+		for _, service := range listener.Services {
+			if service.Name == svc && listener.Port == c.port {
```

Comment:
Realized there was a subtle issue here regarding namespaces, and having two services `space1/foo` and `space2/foo` both being exposed at the same time.

I wonder if it would be more resilient to initialize the `IngressService` we want to create and check equality against that? Something like:
```
exposedSvc := api.IngressService{
			Name:      svc,
			Namespace: svcNamespace,
			Hosts:     c.hosts,
		}
for _, service := range listener.Services {
  if listener.Port == c.port && service == exposedService {
    ...
  }
}
```

Might not be able to use `==` because of the `[]string` for Hosts, but could make a simple `Equals` or `Matches` method I guess?

Reply:
I changed it to just always update the namespace/service combo they passed, and switched this check to look at both service/namespace instead of just service name. I think this is simpler and there wasn't really a reason to block updating an existing service entry anyway.

Reply:
That is also reasonable!